static/apuntes/external-sorts.md

```diff
@@ -0,0 +1,106 @@
+
+Si en cualquier momento el archivo ya no tiene más registros, simplemente obviamos la acción de guardar un nuevo registro en el heap. 
+
+Este algoritmo permite tener particiones de, en promedio, _2C_ registros, teniendo en total _n/2C_ particiones. 
+
+[Aquí](https://www.youtube.com/watch?v=m0v-trXc2yI) un video donde se hace un seguimiento del algoritmo. 
+
+### Opción 3: Natural-Selection sort
+
+Similar al anterior, pero en vez de guardar los archivos en una lista, los guardamos en disco, y nos aseguramos de guardar hasta _C_ registros en ese archivo temporal (y volvemos a leer del archivo, hasta que nos topemos con uno mayor o igual, o tengamos _C_ registros en el archivo temporal). Al escribirlos en disco, permitimos que el heap siga pudiendo guardar registros. Cuando llegamos a _C_ registros en el archivo, simplemente escribimos todo el heap en la partición que estábamos armando, y cargamos los registros del archivo temporal. 
+
+Antes, al toparnos con que teníamos _C_ registros en la lista, no teníamos nada para guardar, pero ahora vamos a tener _C_ registros (los que quedaron en el heap, que siempre va a tener _C_ registros, salvo en el caso final que tendrá uno menos). Entonces, en vez de generar particiones de _2C_, las genera de tamaño _3C_, teniendo entonces _n/3C_ particiones. La contra en este caso es tener que cargar los registros del archivo temporal provenientes de disco. 
+
+### Costo de generar las particiones
+
+En cualquiera de los casos, notar que siempre estaremos trabajando con un heap con _C_ registros, y de alguna forma u otra guardaremos todos los registros en el heap, por lo tanto, el orden de realizar estas operaciones es _O(n log c)_. 
+
+## Costo Total
+
+Teniendo que _k_ valdrá _O(n/C)_ (con alguna constante dividiendo: 1, 2 o 3), tendrémos que nuestro algoritmo finalmente será: 
+
+	SortExterno(n) = O(particiones ordenadas) + O(merge)
+	SortExterno(n) = O(n log c) + O(n log n/c) = O(n log c) + O(n log n) - O(n log c) = O(n log n)
+
+Con eso comprobamos que, lógicamente, nuestro algoritmo ordena en _O(n log n)_. Ahora bien: ¿por qué deberíamos darle importancia si generamos más o menos particiones, si a cuestiones del orden no nos importa? Primero, porque a grandes rasgos es una buena optimización, que se escapa del análisis del orden del algoritmo planteado por la notación _Big Oh_. Pero más importante: estamos hablando de operaciones de lectura y escritura sobre disco. Dichas operaciones toman más tiempo que muchas otras operaciones (si bien ese tiempo sigue siendo constante, es muy grande). Por esto, es que planteamos como posibles mejoras tanto Replacement Selection como Natural Selection. 
+
+## Recomendaciones
+
+A la hora de trabajar con accesos a disco, es importante tener en cuenta que a menos que trabajen con un disco de estado solido (SDD), leer seguido puede ser una operación (muy) costosa, y lo mismo al escribir. 
```

Comment:
Los discos ssd (typo), siguen siendo "lentos" comparados con la memoria y las velocidades máximas se consiguen leyendo el tamaño de bloques del cache del disco (que es "fácil" de abstraer usando lectura secuencial). Los discos PCIe, son 1 órden de magnitud más rápidos, que los deja a 1 órden de magnitud de diferencia de la memoria, que lamentablemente está cada vez más lejos de la velocidad interna del micro (https://www.extremetech.com/computing/268661-cpu-utilization-is-wrong-on-pcs-and-getting-worse-every-year)